File: shared/specs/sdk.ts
// Shiny Fishstick Generated TypeScript SDK
import { chromium, Browser, BrowserContext, Page } from 'playwright';

export class ShinyFishstickSiteSDK {
    private rootUrl: string = "http://localhost:8001";
    private browser: Browser | null = null;
    private context: BrowserContext | null = null;
    private page: Page | null = null;
    private sessionCookies: any[] = [];
    private sessionLocalStorage: any = {};
    private sessionSessionStorage: any = {};

    async start(headless: boolean = true, sessionData?: any): Promise<ShinyFishstickSiteSDK> {
        this.browser = await chromium.launch({ headless });
        this.context = await this.browser.newContext();
        this.page = await this.context.newPage();

        if (sessionData) {
            if (sessionData.cookies) {
                await this.context.addCookies(sessionData.cookies);
            }
            await this.page.goto(this.rootUrl);
            if (sessionData.localStorage) {
                await this.page.evaluate((ls) => {
                    for (let k in ls) localStorage.setItem(k, ls[k]);
                }, sessionData.localStorage);
            }
            if (sessionData.sessionStorage) {
                await this.page.evaluate((ss) => {
                    for (let k in ss) sessionStorage.setItem(k, ss[k]);
                }, sessionData.sessionStorage);
            }
            this.sessionCookies = sessionData.cookies || [];
            this.sessionLocalStorage = sessionData.localStorage || {};
            this.sessionSessionStorage = sessionData.sessionStorage || {};
        } else {
            await this.page.goto(this.rootUrl);
            this.sessionCookies = [];
            this.sessionLocalStorage = {};
            this.sessionSessionStorage = {};
        }
        return this;
    }

    async exportSession(): Promise<any> {
        return {
            cookies: this.sessionCookies || [],
            localStorage: this.sessionLocalStorage || {},
            sessionStorage: this.sessionSessionStorage || {}
        };
    }

    async close(): Promise<void> {
        if (this.browser) await this.browser.close();
    }

    async login(email: string, password: string): Promise<any> {
        
        await this.page.goto(`${this.rootUrl}/login`);
        await this.page.fill("#email", String(email));
        await this.page.fill("#password", String(password));
        await this.page.click("#login-submit-btn");
        await this.page.waitForLoadState("networkidle");
        this.sessionCookies = await this.context.cookies();
        try {
            this.sessionLocalStorage = JSON.parse(await this.page.evaluate(() => JSON.stringify(localStorage)) || '{}');
            this.sessionSessionStorage = JSON.parse(await this.page.evaluate(() => JSON.stringify(sessionStorage)) || '{}');
        } catch (e) {
            this.sessionLocalStorage = {};
            this.sessionSessionStorage = {};
        }
    }

    async search_products(q: string): Promise<any> {
        
        await this.page.goto(`${this.rootUrl}/catalog`);
        await this.page.fill("#search-input", String(q));
        await this.page.click("#search-submit-btn");
        await this.page.waitForLoadState("networkidle");
        this.sessionCookies = await this.context.cookies();
        try {
            this.sessionLocalStorage = JSON.parse(await this.page.evaluate(() => JSON.stringify(localStorage)) || '{}');
            this.sessionSessionStorage = JSON.parse(await this.page.evaluate(() => JSON.stringify(sessionStorage)) || '{}');
        } catch (e) {
            this.sessionLocalStorage = {};
            this.sessionSessionStorage = {};
        }
    }

    async checkout(): Promise<any> {
        
        await this.page.goto(`${this.rootUrl}/checkout`);
        await this.page.click("#checkout-submit-btn");
        await this.page.waitForLoadState("networkidle");
        this.sessionCookies = await this.context.cookies();
        try {
            this.sessionLocalStorage = JSON.parse(await this.page.evaluate(() => JSON.stringify(localStorage)) || '{}');
            this.sessionSessionStorage = JSON.parse(await this.page.evaluate(() => JSON.stringify(sessionStorage)) || '{}');
        } catch (e) {
            this.sessionLocalStorage = {};
            this.sessionSessionStorage = {};
        }
    }

    async add_to_cart(product_id: string, quantity: number): Promise<any> {
        
        const sessionCookie = this.sessionCookies.find(c => c.name === 'session');
        const cookieHeader = sessionCookie ? `session=${sessionCookie.value}` : '';
        const response = await fetch(`${this.rootUrl}/api/cart/add`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Cookie': cookieHeader
            },
            body: JSON.stringify({ "product_id": product_id, "quantity": quantity }),
        });
        return response.json();
    }

}

File: shared/specs/sdk.py
# Shiny Fishstick Generated Python SDK
import requests
import json
from playwright.sync_api import sync_playwright

class ShinyFishstickSiteSDK:
    def __init__(self, root_url="http://localhost:8001"):
        self.root_url = root_url
        self.playwright = None
        self.browser = None
        self.context = None
        self.page = None
        self.session_cookies = []
        self.session_local_storage = {}
        self.session_session_storage = {}

    def start(self, headless=True, session_data=None):
        self.playwright = sync_playwright().start()
        self.browser = self.playwright.chromium.launch(headless=headless)
        self.context = self.browser.new_context()
        self.page = self.context.new_page()

        if session_data:
            cookies = session_data.get("cookies", [])
            if cookies:
                self.context.add_cookies(cookies)
            self.page.goto(self.root_url)

            ls = session_data.get("localStorage", {})
            ss = session_data.get("sessionStorage", {})
            if ls:
                self.page.evaluate("ls => { for (let k in ls) { localStorage.setItem(k, ls[k]); } }", ls)
            if ss:
                self.page.evaluate("ss => { for (let k in ss) { sessionStorage.setItem(k, ss[k]); } }", ss)
            self.session_cookies = cookies
            self.session_local_storage = ls
            self.session_session_storage = ss
        else:
            self.page.goto(self.root_url)
            self.session_cookies = []
            self.session_local_storage = {}
            self.session_session_storage = {}
        return self

    def export_session(self):
        return {
            "cookies": self.session_cookies or [],
            "localStorage": self.session_local_storage or {},
            "sessionStorage": self.session_session_storage or {}
        }

    def close(self):
        if self.browser:
            self.browser.close()
        if self.playwright:
            self.playwright.stop()

    def login(self, email, password):
        """Logs in the user with credentials"""

        self.page.goto(self.root_url + "/login")
        self.page.fill("#email", str(email))
        self.page.fill("#password", str(password))
        self.page.click("#login-submit-btn")
        self.page.wait_for_load_state("networkidle")
        self.session_cookies = self.page.context.cookies()
        import json
        try:
            ls_str = self.page.evaluate("() => JSON.stringify(localStorage)")
            self.session_local_storage = json.loads(ls_str or '{}')
            ss_str = self.page.evaluate("() => JSON.stringify(sessionStorage)")
            self.session_session_storage = json.loads(ss_str or '{}')
        except Exception:
            self.session_local_storage = {}
            self.session_session_storage = {}

    def search_products(self, q):
        """Searches for products in the store"""

        self.page.goto(self.root_url + "/catalog")
        self.page.fill("#search-input", str(q))
        self.page.click("#search-submit-btn")
        self.page.wait_for_load_state("networkidle")
        self.session_cookies = self.page.context.cookies()
        import json
        try:
            ls_str = self.page.evaluate("() => JSON.stringify(localStorage)")
            self.session_local_storage = json.loads(ls_str or '{}')
            ss_str = self.page.evaluate("() => JSON.stringify(sessionStorage)")
            self.session_session_storage = json.loads(ss_str or '{}')
        except Exception:
            self.session_local_storage = {}
            self.session_session_storage = {}

    def checkout(self):
        """Proceeds to place the order and checkout"""

        self.page.goto(self.root_url + "/checkout")
        self.page.click("#checkout-submit-btn")
        self.page.wait_for_load_state("networkidle")
        self.session_cookies = self.page.context.cookies()
        import json
        try:
            ls_str = self.page.evaluate("() => JSON.stringify(localStorage)")
            self.session_local_storage = json.loads(ls_str or '{}')
            ss_str = self.page.evaluate("() => JSON.stringify(sessionStorage)")
            self.session_session_storage = json.loads(ss_str or '{}')
        except Exception:
            self.session_local_storage = {}
            self.session_session_storage = {}

    def add_to_cart(self, product_id, quantity=1):
        """Adds the current product to the shopping cart"""

        # API Action execution
        session_val = None
        if self.session_cookies:
            session_val = next((c["value"] for c in self.session_cookies if c["name"] == "session"), None)
        cookies = {}
        if session_val:
            cookies = {"session": session_val}
        payload = {
        "product_id": product_id,
        "quantity": quantity
    }
        query_params = {}
        headers = {}
        res = requests.post(
            self.root_url + "/api/cart/add", json=payload,
            cookies=cookies,
            headers=headers
        )
        return res.json()

